cmd/subctl/upgrade.go

```diff
@@ -0,0 +1,271 @@
+		// Already running the right version
+		return nil, nil //nolint:nilnil // A nil command means subctl shouldn't be restarted.
+	}
+
+	if to != "" {
+		if len(to) > 1 && to[0:1] == "v" {
+			to = to[1:]
+		}
+
+		toVersion, err := semver.NewVersion(to)
+		if toVersion == nil {
+			return nil, status.Error(err, "Invalid target version")
+		}
+
+		// semver needs a dotted triplet, which is at least five characters;
+		// on development or unknown versions, assume we need to upgrade
+		if len(version.Version) >= 5 && version.Version[0:5] != "devel" {
+			currentVersion, err := semver.NewVersion(version.Version)
+			if currentVersion == nil {
+				return nil, status.Error(err, "Error parsing current subctl version")
+			}
+
+			if toVersion.LessThan(*currentVersion) || toVersion.Equal(*currentVersion) {
+				return nil, nil //nolint:nilnil // A nil command means subctl shouldn't be restarted.
+			}
+		}
+
+		targetVersionString = "v" + toVersion.String()
+	}
+
+	status.Start(fmt.Sprintf("Upgrading subctl from %s to %s, replacing %s", version.Version, targetVersionString, os.Args[0]))
+	defer status.End()
+
+	url := "https://get.submariner.io"
+
+	absolutePath, err := filepath.Abs(os.Args[0])
+	if err != nil {
+		return nil, status.Error(err, "Error determining the installation path")
+	}
+
+	_, err = exec.Command( //nolint:gosec // The user-controlled variables are sanitised above
+		"sh", "-c", "curl "+url+" | VERSION="+targetVersionString+" DESTDIR="+filepath.Dir(absolutePath)+" bash").CombinedOutput()
+	if err != nil {
+		return nil, status.Error(err, "Error upgrading subctl")
+	}
+
+	status.Success("Upgraded and installed subctl version: %s", to)
+
+	return &absolutePath, nil
+}
+
+func upgradeSubmariner(clusterInfo *cluster.Info, _ string, status reporter.Interface) error {
+	ctx := context.TODO()
+
+	status.Start("Starting upgrade process")
+	defer status.End()
+
+	brokerObj, brokerFound, err := getBroker(clusterInfo.RestConfig, constants.DefaultBrokerNamespace)
+	if err != nil {
+		return err
+	}
+
+	if brokerFound {
+		// Role updates are part of Broker redeploy
+		err = upgradeBroker(ctx, clusterInfo, status, brokerObj.Namespace, brokerObj.Spec)
+		if err != nil {
+			return status.Error(err, "Error upgrading Broker")
+		}
+	}
+
+	var repository string
+	var debug bool
+	var imageOverride map[string]string
+
+	if clusterInfo.Submariner != nil {
+		repository = clusterInfo.Submariner.Spec.Repository
+		imageOverride = clusterInfo.Submariner.Spec.ImageOverrides
+		debug = clusterInfo.Submariner.Spec.Debug
+	} else if clusterInfo.ServiceDiscovery != nil {
+		repository = clusterInfo.ServiceDiscovery.Spec.Repository
+		imageOverride = clusterInfo.ServiceDiscovery.Spec.ImageOverrides
+		debug = clusterInfo.ServiceDiscovery.Spec.Debug
+	}
+
+	// Upgrade Operator if deployed
+	if err := upgradeOperator(ctx, clusterInfo, repository, debug, imageOverride, status); err != nil {
+		return status.Error(err, "Error upgrading Operator")
```

Comment:
The status should be updated by `upgradeOperator` since it starts it.

```suggestion
		return err
```